src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -3465,45 +3465,80 @@ class ArnoldRenderer final : public ArnoldRendererBase
 
 		ArnoldRenderer( RenderType renderType, const std::string &fileName, const IECore::MessageHandlerPtr &messageHandler )
 			:	ArnoldRendererBase( nodeDeleter( renderType ) ),
-				m_globals( new ArnoldGlobals( renderType, fileName, m_shaderCache.get() ) )
+				m_messageHandler( messageHandler ), m_globals( new ArnoldGlobals( renderType, fileName, m_shaderCache.get() ) )
 		{
+			if( m_messageHandler )
+			{
+				AiMsgSetCallback( &ArnoldRenderer::aiMsgCallback );
+			}
 		}
 
 		~ArnoldRenderer() override
 		{
 			pause();
+			if( g_currentMessageHandler == m_messageHandler )
+			{
+				g_currentMessageHandler = nullptr;
+			}
 		}
 
 		void option( const IECore::InternedString &name, const IECore::Object *value ) override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->option( name, value );
 		}
 
 		void output( const IECore::InternedString &name, const IECoreScene::Output *output ) override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->output( name, output );
 		}
 
 		ObjectInterfacePtr camera( const std::string &name, const IECoreScene::Camera *camera, const AttributesInterface *attributes ) override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->camera( name, camera );
 			return ArnoldRendererBase::camera( name, camera, attributes );
 		}
 
 		void render() override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
+			// Promote our handler to global
+			if( m_messageHandler )
+			{
+				g_currentMessageHandler = m_messageHandler;
+			}
+
 			m_shaderCache->clearUnused();
 			m_instanceCache->clearUnused();
 			m_globals->render();
 		}
 
 		void pause() override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->pause();
 		}
 
 	private :
 
+		// Arnold's singleton-centric design means there is no way to pass any
+		// instance-specific references as part of the log callback. As such
+		// we have to make do with a global handler.
+
+		static void aiMsgCallback( int logmask, int severity, const char *msg_string, int tabs )
+		{
+			if( g_currentMessageHandler )
```

Comment:
I think we'd do better to avoid this check, by only installing our callback when we have a handler.

Reply:
We do only install the handler in that case, I guess I was just wanting to safeguard against the fact that this was technically a global...

Reply:
Refactored in https://github.com/GafferHQ/gaffer/pull/3767/commits/d917f02a78e2c4a5bc1f11efa98d8fed04f9c996